R-package/tests/testthat/test_basic.R

```diff
@@ -2042,3 +2042,147 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
   expect_equal(pred1, pred2)
 
 })
+
+context("monotone constraints")
+
+.generate_trainset_for_monotone_constraints_tests <- function(x3_to_categorical) {
+  n_samples <- 3000L
+  x1_positively_correlated_with_y <- rnorm(n = n_samples)
+  x2_negatively_correlated_with_y <- rnorm(n = n_samples)
+  x3_negatively_correlated_with_y <- rnorm(n = n_samples)
+  if (x3_to_categorical) {
+    x3_negatively_correlated_with_y <- as.integer(abs(runif(n_samples) / 0.25))
+    categorical_features <- 3L
+  } else {
+    x3_negatively_correlated_with_y <- runif(n_samples)
+    categorical_features <- NULL
+  }
+  X <- data.matrix(
+    data.frame(
+      list(
+        x1_positively_correlated_with_y
+        , x2_negatively_correlated_with_y
+        , x3_negatively_correlated_with_y
+      )
+    )
+  )
+  zs <- rnorm(n = n_samples, mean = 0.0, sd = 0.01)
+  scales <- 10.0 * rnorm(6L + 0.5)
```

Comment:
Different bounds compared to Python original test.

Reply:
ah I see, my parentheses were wrong

https://github.com/microsoft/LightGBM/blob/53ffba7c5c4a4d64853ad6c5409735b986283d2d/tests/python_package_test/test_engine.py#L1240

Reply:
fixed in https://github.com/microsoft/LightGBM/commit/e7b2048a6f8b9e42e5dda4b9012d8c9dcc06a9d7

Reply:
Still different bounds. I guess you need `runif(n = 6L, min = 0.0, max = 1.0)` to mimic `np.random.random(6)` part.

Reply:
fixed in https://github.com/microsoft/LightGBM/pull/4352/commits/a15596a5b4d44f8c0ac9e8ab5069c73a28669924